File: backend/app/open_node/domain/changes.py
from datetime import datetime
from enum import StrEnum
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from open_node.domain.inventory import AgentCommandCreate, AgentCommandRead


class AgentChangeSetStatus(StrEnum):
    PLANNED = "planned"
    DISPATCHED = "dispatched"
    ROLLBACK_QUEUED = "rollback_queued"


class AgentChangeSetStepCreate(BaseModel):
    server_id: UUID
    label: str = Field(default="", max_length=160)
    forward: AgentCommandCreate
    rollback: AgentCommandCreate | None = None

    @field_validator("label")
    @classmethod
    def normalize_label(cls, value: str) -> str:
        return value.strip()


class AgentChangeSetCreate(BaseModel):
    name: str = Field(min_length=1, max_length=160)
    description: str = Field(default="", max_length=1000)
    rollback_on_failure: bool = True
    dispatch: bool = False
    steps: list[AgentChangeSetStepCreate] = Field(min_length=1, max_length=100)

    @field_validator("name")
    @classmethod
    def normalize_name(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("name is empty")
        return normalized

    @field_validator("description")
    @classmethod
    def normalize_description(cls, value: str) -> str:
        return value.strip()


class AgentChangeSetRollbackRequest(BaseModel):
    reason: str = Field(default="", max_length=1000)

    @field_validator("reason")
    @classmethod
    def normalize_reason(cls, value: str) -> str:
        return value.strip()


class AgentChangeSetStepRead(BaseModel):
    id: UUID
    change_set_id: UUID
    sequence: int
    server_id: UUID
    label: str
    forward: AgentCommandCreate
    rollback: AgentCommandCreate | None = None
    forward_command: AgentCommandRead | None = None
    rollback_command: AgentCommandRead | None = None
    created_at: datetime
    updated_at: datetime


class AgentChangeSetRead(BaseModel):
    id: UUID
    name: str
    description: str
    status: AgentChangeSetStatus
    rollback_on_failure: bool
    rollback_reason: str = ""
    steps: list[AgentChangeSetStepRead]
    created_at: datetime
    updated_at: datetime


class AgentChangeSetsResponse(BaseModel):
    change_sets: list[AgentChangeSetRead]
    license_required: Literal[False] = False


class AgentChangeSetResponse(BaseModel):
    change_set: AgentChangeSetRead
    commands: list[AgentCommandRead] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    license_required: Literal[False] = False

File: frontend/src/domain/changes.ts
import type { AgentCommand, AgentCommandCreateRequest } from "./inventory";

export type AgentChangeSetStatus = "planned" | "dispatched" | "rollback_queued" | "succeeded"
  | "failed" | "rolled_back" | "rollback_failed" | "rollback_incomplete" | "cancelled"
  | "accepted" | "needs_review";

export interface AgentChangeSetStepCreateRequest {
  server_id: string;
  label?: string;
  forward: AgentCommandCreateRequest;
  rollback?: AgentCommandCreateRequest | null;
}

export interface AgentChangeSetCreateRequest {
  name: string;
  description?: string;
  rollback_on_failure?: boolean;
  dispatch?: boolean;
  steps: AgentChangeSetStepCreateRequest[];
}

export interface AgentRoutedOutboundChangeSetCreateRequest {
  server_id: string;
  inbound_tag: string;
  inbound_protocol?: string;
  label: string;
  outbound: Record<string, unknown>;
  parent_ref?: string | null;
  admin_username?: string;
  admin_email?: string | null;
  outbound_tag?: string | null;
  marktag?: string | null;
  node_name?: string | null;
  client?: Record<string, unknown> | null;
  sniffing_exclude_domains?: string[];
  add_reality_sniffing_excludes?: boolean;
  command_timeout_ms?: number;
  rollback_on_failure?: boolean;
  dispatch?: boolean;
}

export interface AgentChangeSetRollbackRequest {
  reason?: string;
}

export interface AgentChangeSetStep {
  archived?: boolean;
  server_name?: string | null;
  id: string;
  change_set_id: string;
  sequence: number;
  server_id: string;
  label: string;
  forward: AgentCommandCreateRequest;
  rollback?: AgentCommandCreateRequest | null;
  forward_command?: AgentCommand | null;
  rollback_command?: AgentCommand | null;
  rollback_history?: AgentCommand[];
  created_at: string;
  updated_at: string;
}

export interface AgentChangeSet {
  id: string;
  name: string;
  description: string;
  status: AgentChangeSetStatus;
  rollback_on_failure: boolean;
  rollback_reason: string;
  resolution_reason?: string;
  held_server_ids?: string[];
  blocking_command_ids?: string[];
  warnings?: string[];
  steps: AgentChangeSetStep[];
  created_at: string;
  updated_at: string;
}

export interface AgentChangeSetsResponse {
  change_sets: AgentChangeSet[];
  license_required: false;
}

export interface AgentChangeSetResponse {
  change_set: AgentChangeSet;
  commands: AgentCommand[];
  warnings: string[];
  license_required: false;
}

export function changeSetActions(change: AgentChangeSet | null) {
  const status = change?.steps?.some(step => step.archived) ? undefined : change?.status;
  return {
    dispatch: status === "planned",
    rollback: !!status && ["planned", "dispatched", "succeeded", "failed", "rollback_failed", "rollback_incomplete"].includes(status),
    accept: !!status && ["failed", "rollback_failed", "rollback_incomplete", "needs_review"].includes(status)
      && !change?.blocking_command_ids?.length,
    retry: status === "rollback_failed",
  };
}
